README.md
# dynamic_choice_chip

This flutter plugin is for dynamic choice chip

## Getting Started

Using this plugin developer can render the choice chip dynamically.
Also, developer can make the app more scalable using this plugin, and it saves time too.

In case of any issue or any customization in the plugin you can reach me out on barsainya.satyam.sb@gmail.com

CHANGELOG.md
## 0.0.1

* This release has whole lot of customization which a developer requires in the choice chip
* Future releases coming soon with more customized options
* This plugin has not been tested in iOS so it might have some issue


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:dynamic_choice_chip_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/dynamic_choice_chip_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const MethodChannel channel = MethodChannel('dynamic_choice_chip');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });
}


example/lib/main.dart
import 'package:dynamic_choice_chip/dynamic_choice_chip.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MaterialApp(debugShowCheckedModeBanner: false, home: MyApp()));
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  MyAppState createState() => MyAppState();
}

class MyAppState extends State<MyApp> {
  List dataSource = ["India", "USA", "UK", "Indie"];
  Map countryMap = {};
  String countryName = "";

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Scaffold(
        backgroundColor: Colors.blueGrey,
        body: Padding(
          padding: const EdgeInsets.all(8.0),
          child: Column(
            children: [
              ElevatedButton(
                onPressed: () {
                  showModalBottomSheet(
                      backgroundColor: Colors.teal,
                      shape: RoundedRectangleBorder(
                        borderRadius: BorderRadius.circular(15.0),
                      ),
                      context: context,
                      builder: (context) {
                        // return DynamicMultiChoiceChip(
                        //   activeWidgetColor: Colors.blue,
                        //   unactiveWidgetColor: Colors.pink,
                        //   cancelButtonBgColor: Colors.blue,
                        //   okButtonBgColor: Colors.blue,
                        //   prefixIconColor: Colors.white,
                        //   prefixIcon: Icons.search,
                        //   textFormFieldLabelText: "Search Country",
                        //   textFormFieldBorderColor: Colors.white,
                        //   renderedWidgetBorderColor: Colors.white,
                        //   dataSource: dataSource,
                        //   renderedTextColor: Colors.white,
                        //   renderedTextSize: 12,
                        //   onTap: (values) {
                        //     setState(() {
                        //       countryMap = values;
                        //       values.forEach((key, value) {
                        //         print('Key = $key : Value = $value');
                        //       });
                        //     });
                        //
                        //   },
                        // );
                        return DynamicChoiceChip(
                          activeWidgetColor: Colors.blue,
                          unactiveWidgetColor: Colors.pink,
                          cancelButtonBgColor: Colors.blue,
                          okButtonBgColor: Colors.blue,
                          prefixIconColor: Colors.white,
                          prefixIcon: Icons.search,
                          textFormFieldLabelText: "Search Country",
                          textFormFieldBorderColor: Colors.white,
                          renderedWidgetBorderColor: Colors.white,
                          dataSource: dataSource,
                          renderedTextColor: Colors.white,
                          renderedTextSize: 12,
                          onTap: (index, value) {
                            setState(() {
                              countryName = value;
                              // print("index : ${index} : value : ${value}");
                            });
                          },
                        );
                      });
                  // setState(() {});
                },
                child: const Text("Click"),
              ),
              // ListView.builder(
              //     shrinkWrap: true,
              //     physics: const ClampingScrollPhysics(),
              //     itemCount: countryMap.length,
              //     itemBuilder: (BuildContext context, int index) {
              //       return Padding(
              //         padding: const EdgeInsets.all(8.0),
              //         child: Container(
              //           color: Colors.blue,
              //           child: Padding(
              //             padding: const EdgeInsets.all(8.0),
              //             child: Center(child: Text(countryMap[index])),
              //           ),
              //         ),
              //       );
              //     })
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: Container(
                  color: Colors.blue,
                  child: Padding(
                    padding: const EdgeInsets.all(8.0),
                    child: Center(child: Text(countryName)),
                  ),
                ),
              )
            ],
          ),
        ),
      ),
    );
  }
}


lib/dynamic_choice_chip.dart
import 'package:assorted_layout_widgets/assorted_layout_widgets.dart';
import 'package:flutter/material.dart';

//ignore: must_be_immutable
class DynamicChoiceChip extends StatefulWidget {
  final IconData prefixIcon;
  final Color prefixIconColor;
  final String textFormFieldLabelText;
  final Color textFormFieldBorderColor;
  final Color renderedWidgetBorderColor;
  final Color renderedTextColor;
  final double renderedTextSize;
  final List dataSource;
  final Color okButtonBgColor;
  final Color cancelButtonBgColor;
  final Color activeWidgetColor;
  final Color unactiveWidgetColor;
  final Function(int, String) onTap;

  const DynamicChoiceChip(
      {Key? key,
      required this.prefixIcon,
      required this.prefixIconColor,
      required this.textFormFieldLabelText,
      required this.textFormFieldBorderColor,
      required this.renderedWidgetBorderColor,
      required this.dataSource,
      required this.renderedTextColor,
      required this.renderedTextSize,
      required this.okButtonBgColor,
      required this.cancelButtonBgColor,
      required this.activeWidgetColor,
      required this.unactiveWidgetColor,
      required this.onTap})
      : super(key: key);

  @override
  DynamicChoiceChipState createState() => DynamicChoiceChipState();
}

class DynamicChoiceChipState extends State<DynamicChoiceChip> {
  String selectedIndex = "";
  List filteredDataList = [];
  late int originalIndex;

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        Container(
          width: MediaQuery.of(context).size.width,
          height: 90,
          decoration: BoxDecoration(
            color: Colors.transparent,
            borderRadius: BorderRadius.circular(15),
          ),
          child: Padding(
            padding: const EdgeInsets.all(10.0),
            child: TextFormField(
              onChanged: (text) {
                List filteredList = [];
                for (int i = 0; i < widget.dataSource.length; i++) {
                  if ((widget.dataSource[i].toLowerCase())
                      .toString()
                      .contains(text)) {
                    setState(() {
                      filteredList.add(widget.dataSource[i]);
                      originalIndex = i;
                      filteredDataList = filteredList.toSet().toList();
                    });
                  }
                }
              },
              decoration: InputDecoration(
                  prefixIcon: Icon(
                    widget.prefixIcon,
                    color: widget.prefixIconColor,
                  ),
                  // fillColor: Colors.green,
                  labelStyle: const TextStyle(color: Colors.white),
                  border: InputBorder.none,
                  labelText: widget.textFormFieldLabelText,
                  enabledBorder: OutlineInputBorder(
                    borderRadius: const BorderRadius.all(
                      Radius.circular(15.0),
                    ),
                    borderSide: BorderSide(
                        width: 3, color: widget.textFormFieldBorderColor),
                    // borderRadius: BorderRadius.circular(15),
                  )),
            ),
          ),
        ),
        // Container(child: Text()),
        filteredDataList.isNotEmpty
            ? Expanded(
                child: SingleChildScrollView(
                  child: WrapSuper(
                      wrapFit: WrapFit.larger,
                      alignment: WrapSuperAlignment.left,
                      wrapType: WrapType.fit,
                      spacing: 5.0,
                      lineSpacing: 2.0,
                      children: List.generate(filteredDataList.length, (index) {
                        return GestureDetector(
                          onTap: () {
                            setState(() {
                              selectedIndex = index.toString();
                            });
                          },
                          child: Padding(
                            padding: const EdgeInsets.all(8.0),
                            child: Container(
                              height: 30,
                              decoration: BoxDecoration(
                                color: selectedIndex == index.toString()
                                    ? widget.activeWidgetColor
                                    : widget.unactiveWidgetColor,
                                border: Border.all(
                                    color: widget.renderedWidgetBorderColor),
                                borderRadius: BorderRadius.circular(12),
                              ),
                              child: Center(
                                child: Padding(
                                  padding: const EdgeInsets.all(8.0),
                                  child: Text(
                                    filteredDataList[index],
                                    // maxLines: 1,
                                    style: TextStyle(
                                        color: widget.renderedTextColor,
                                        fontSize: widget.renderedTextSize),
                                  ),
                                ),
                              ),
                            ),
                          ),
                        );
                      })),
                ),
              )
            : Expanded(
                child: SingleChildScrollView(
                  child: WrapSuper(
                      wrapFit: WrapFit.larger,
                      alignment: WrapSuperAlignment.left,
                      wrapType: WrapType.fit,
                      spacing: 5.0,
                      lineSpacing: 2.0,
                      children:
                          List.generate(widget.dataSource.length, (index) {
                        return GestureDetector(
                          onTap: () {
                            setState(() {
                              selectedIndex = index.toString();
                            });
                          },
                          child: Padding(
                            padding: const EdgeInsets.all(8.0),
                            child: Container(
                              height: 30,
                              decoration: BoxDecoration(
                                color: selectedIndex == index.toString()
                                    ? widget.activeWidgetColor
                                    : widget.unactiveWidgetColor,
                                border: Border.all(
                                    color: widget.renderedWidgetBorderColor),
                                borderRadius: BorderRadius.circular(12),
                              ),
                              child: Center(
                                child: Padding(
                                  padding: const EdgeInsets.all(8.0),
                                  child: Text(
                                    widget.dataSource[index],
                                    // maxLines: 1,
                                    style: TextStyle(
                                        color: widget.renderedTextColor,
                                        fontSize: widget.renderedTextSize),
                                  ),
                                ),
                              ),
                            ),
                          ),
                        );
                      })),
                ),
              ),
        filteredDataList.isNotEmpty
            ? ButtonBarSuper(
                wrapType: WrapType.fit,
                wrapFit: WrapFit.larger,
                spacing: 2.0,
                // Horizontal spacing between buttons.
                lineSpacing: 10.0,
                // Vertical spacing between button lines.
                buttonHeight: 48,
                buttonMinWidth: 40,
                children: [
                  ElevatedButton(
                      style: ButtonStyle(
                        backgroundColor: MaterialStateProperty.all<Color>(
                            widget.cancelButtonBgColor),
                        shape:
                            MaterialStateProperty.all<RoundedRectangleBorder>(
                          RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(15.0),
                          ),
                        ),
                      ),
                      onPressed: () {
                        Navigator.of(context).pop();
                      },
                      child: const Text("Cancel")),
                  ElevatedButton(
                      style: ButtonStyle(
                        backgroundColor: MaterialStateProperty.all<Color>(
                            widget.okButtonBgColor),
                        shape:
                            MaterialStateProperty.all<RoundedRectangleBorder>(
                          RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(15.0),
                          ),
                        ),
                      ),
                      onPressed: () {
                        widget.onTap(originalIndex,
                            filteredDataList[int.parse(selectedIndex)]);
                        Navigator.pop(context);
                      },
                      child: const Text("Ok"))
                ],
              )
            : ButtonBarSuper(
                wrapType: WrapType.fit,
                wrapFit: WrapFit.larger,
                spacing: 2.0,
                // Horizontal spacing between buttons.
                lineSpacing: 10.0,
                // Vertical spacing between button lines.
                buttonHeight: 48,
                buttonMinWidth: 40,
                children: [
                  ElevatedButton(
                      style: ButtonStyle(
                        backgroundColor: MaterialStateProperty.all<Color>(
                            widget.cancelButtonBgColor),
                        shape:
                            MaterialStateProperty.all<RoundedRectangleBorder>(
                          RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(15.0),
                          ),
                        ),
                      ),
                      onPressed: () {
                        Navigator.of(context).pop();
                      },
                      child: const Text("Cancel")),
                  ElevatedButton(
                      style: ButtonStyle(
                        backgroundColor: MaterialStateProperty.all<Color>(
                            widget.okButtonBgColor),
                        shape:
                            MaterialStateProperty.all<RoundedRectangleBorder>(
                          RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(15.0),
                          ),
                        ),
                      ),
                      onPressed: () {
                        widget.onTap(int.parse(selectedIndex),
                            widget.dataSource[int.parse(selectedIndex)]);
                        Navigator.pop(context);
                      },
                      child: const Text("Ok"))
                ],
              )
      ],
    );
  }
}


android/src/main/java/com/example/dynamic_choice_chip/DynamicChoiceChipPlugin.java
package com.example.dynamic_choice_chip;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/** DynamicChoiceChipPlugin */
public class DynamicChoiceChipPlugin implements FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "dynamic_choice_chip");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}


lib/dynamic_multi_choice_chip.dart
import 'package:assorted_layout_widgets/assorted_layout_widgets.dart';
import 'package:flutter/material.dart';

//ignore: must_be_immutable
class DynamicMultiChoiceChip extends StatefulWidget {
  final IconData prefixIcon;
  final Color prefixIconColor;
  final String textFormFieldLabelText;
  final Color textFormFieldBorderColor;
  final Color renderedWidgetBorderColor;
  final Color renderedTextColor;
  final double renderedTextSize;
  final List dataSource;
  final Color okButtonBgColor;
  final Color cancelButtonBgColor;
  final Color activeWidgetColor;
  final Color unactiveWidgetColor;
  final Function(Map<int, String>) onTap;

  const DynamicMultiChoiceChip(
      {Key? key,
      required this.prefixIcon,
      required this.prefixIconColor,
      required this.textFormFieldLabelText,
      required this.textFormFieldBorderColor,
      required this.renderedWidgetBorderColor,
      required this.dataSource,
      required this.renderedTextColor,
      required this.renderedTextSize,
      required this.okButtonBgColor,
      required this.cancelButtonBgColor,
      required this.activeWidgetColor,
      required this.unactiveWidgetColor,
      required this.onTap})
      : super(key: key);

  @override
  DynamicMultiChoiceChipState createState() => DynamicMultiChoiceChipState();
}

class DynamicMultiChoiceChipState extends State<DynamicMultiChoiceChip> {
  List<int> selectedItems = [];
  List<Map<int, String>> selectedItemsMap = [];
  Map<int, String> items = {};
  List filteredDataList = [];

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        Container(
          width: MediaQuery.of(context).size.width,
          height: 90,
          decoration: BoxDecoration(
            color: Colors.transparent,
            borderRadius: BorderRadius.circular(15),
          ),
          child: Padding(
            padding: const EdgeInsets.all(10.0),
            child: TextFormField(
              onChanged: (text) {
                List filteredList = [];
                for (int i = 0; i < widget.dataSource.length; i++) {
                  if ((widget.dataSource[i].toLowerCase())
                      .toString()
                      .contains(text)) {
                    setState(() {
                      filteredList.add(widget.dataSource[i]);
                      filteredDataList = filteredList.toSet().toList();
                    });
                  }
                }
              },
              decoration: InputDecoration(
                  prefixIcon: Icon(
                    widget.prefixIcon,
                    color: widget.prefixIconColor,
                  ),
                  // fillColor: Colors.green,
                  labelStyle: const TextStyle(color: Colors.white),
                  border: InputBorder.none,
                  labelText: widget.textFormFieldLabelText,
                  enabledBorder: OutlineInputBorder(
                    borderRadius: const BorderRadius.all(
                      Radius.circular(15.0),
                    ),
                    borderSide: BorderSide(
                        width: 3, color: widget.textFormFieldBorderColor),
                    // borderRadius: BorderRadius.circular(15),
                  )),
            ),
          ),
        ),
        filteredDataList.isNotEmpty
            ? Expanded(
                child: SingleChildScrollView(
                  child: WrapSuper(
                      wrapFit: WrapFit.larger,
                      alignment: WrapSuperAlignment.left,
                      wrapType: WrapType.fit,
                      spacing: 5.0,
                      lineSpacing: 2.0,
                      children: List.generate(filteredDataList.length, (index) {
                        return GestureDetector(
                          onTap: () {
                            setState(() {
                              if (selectedItems.contains(index)) {
                                selectedItems.remove(index);
                              } else {
                                selectedItems.add(index);
                              }
                            });
                          },
                          child: Padding(
                            padding: const EdgeInsets.all(8.0),
                            child: Container(
                              height: 30,
                              decoration: BoxDecoration(
                                color: selectedItems.contains(index)
                                    ? widget.activeWidgetColor
                                    : widget.unactiveWidgetColor,
                                border: Border.all(
                                    color: widget.renderedWidgetBorderColor),
                                borderRadius: BorderRadius.circular(12),
                              ),
                              child: Center(
                                child: Padding(
                                  padding: const EdgeInsets.all(8.0),
                                  child: Text(
                                    filteredDataList[index],
                                    // maxLines: 1,
                                    style: TextStyle(
                                        color: widget.renderedTextColor,
                                        fontSize: widget.renderedTextSize),
                                  ),
                                ),
                              ),
                            ),
                          ),
                        );
                      })),
                ),
              )
            : Expanded(
                child: SingleChildScrollView(
                  child: WrapSuper(
                      wrapFit: WrapFit.larger,
                      alignment: WrapSuperAlignment.left,
                      wrapType: WrapType.fit,
                      spacing: 5.0,
                      lineSpacing: 2.0,
                      children:
                          List.generate(widget.dataSource.length, (index) {
                        return GestureDetector(
                          onTap: () {
                            setState(() {
                              if (selectedItems.contains(index)) {
                                selectedItems.remove(index);
                              } else {
                                selectedItems.add(index);
                              }
                            });
                          },
                          child: Padding(
                            padding: const EdgeInsets.all(8.0),
                            child: Container(
                              height: 30,
                              decoration: BoxDecoration(
                                color: selectedItems.contains(index)
                                    ? widget.activeWidgetColor
                                    : widget.unactiveWidgetColor,
                                border: Border.all(
                                    color: widget.renderedWidgetBorderColor),
                                borderRadius: BorderRadius.circular(12),
                              ),
                              child: Center(
                                child: Padding(
                                  padding: const EdgeInsets.all(8.0),
                                  child: Text(
                                    widget.dataSource[index],
                                    // maxLines: 1,
                                    style: TextStyle(
                                        color: widget.renderedTextColor,
                                        fontSize: widget.renderedTextSize),
                                  ),
                                ),
                              ),
                            ),
                          ),
                        );
                      })),
                ),
              ),
        filteredDataList.isNotEmpty
            ? ButtonBarSuper(
                wrapType: WrapType.fit,
                wrapFit: WrapFit.larger,
                spacing: 2.0,
                // Horizontal spacing between buttons.
                lineSpacing: 10.0,
                // Vertical spacing between button lines.
                buttonHeight: 48,
                buttonMinWidth: 40,
                children: [
                  ElevatedButton(
                      style: ButtonStyle(
                        backgroundColor: MaterialStateProperty.all<Color>(
                            widget.cancelButtonBgColor),
                        shape:
                            MaterialStateProperty.all<RoundedRectangleBorder>(
                          RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(15.0),
                          ),
                        ),
                      ),
                      onPressed: () {
                        Navigator.of(context).pop();
                      },
                      child: const Text("Cancel")),
                  ElevatedButton(
                      style: ButtonStyle(
                        backgroundColor: MaterialStateProperty.all<Color>(
                            widget.okButtonBgColor),
                        shape:
                            MaterialStateProperty.all<RoundedRectangleBorder>(
                          RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(15.0),
                          ),
                        ),
                      ),
                      onPressed: () {
                        for (int i = 0; i < selectedItems.length; i++) {
                          items.putIfAbsent(i,
                              () => filteredDataList[selectedItems[i]]);
                          // selectedItemsMap.add(items);
                        }
                        widget.onTap(items);
                        Navigator.pop(context);
                      },
                      child: const Text("Ok"))
                ],
              )
            : ButtonBarSuper(
                wrapType: WrapType.fit,
                wrapFit: WrapFit.larger,
                spacing: 2.0,
                // Horizontal spacing between buttons.
                lineSpacing: 10.0,
                // Vertical spacing between button lines.
                buttonHeight: 48,
                buttonMinWidth: 40,
                children: [
                  ElevatedButton(
                      style: ButtonStyle(
                        backgroundColor: MaterialStateProperty.all<Color>(
                            widget.cancelButtonBgColor),
                        shape:
                            MaterialStateProperty.all<RoundedRectangleBorder>(
                          RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(15.0),
                          ),
                        ),
                      ),
                      onPressed: () {
                        Navigator.of(context).pop();
                      },
                      child: const Text("Cancel")),
                  ElevatedButton(
                      style: ButtonStyle(
                        backgroundColor: MaterialStateProperty.all<Color>(
                            widget.okButtonBgColor),
                        shape:
                            MaterialStateProperty.all<RoundedRectangleBorder>(
                          RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(15.0),
                          ),
                        ),
                      ),
                      onPressed: () {
                        for (int i = 0; i < selectedItems.length; i++) {
                          items.putIfAbsent(i,
                              () => widget.dataSource[selectedItems[i]]);
                        }
                        widget.onTap(items);
                        Navigator.pop(context);
                      },
                      child: const Text("Ok"))
                ],
              )
      ],
    );
  }
}
